src/uct/ib/rdmacm/rdmacm_cm.h

```diff
@@ -50,6 +56,36 @@ typedef struct uct_rdmacm_cm_config {
     double          timeout;
 } uct_rdmacm_cm_config_t;
 
+struct uct_rdmacm_cm_reserved_qpn_blk {
+    ucs_bitmap_t(256)       bmp;
+    uint32_t                first_qpn;
+    unsigned int            next_avail_qpn_offset;
```

Comment:
why do we need this field? can't we just search for first available bit in the bitmap?

Reply:
@yosefe No, maybe we can't do it. The bit which was used and freed can't be treated as available bit again. This is referred to rdma-core code. Please see the same comments also exist in rdma-core as following.
https://github.com/leibin2014/ucx/blob/28324b51eb86a1ec8d07381d573cfe515efa6f4a/src/uct/ib/rdmacm/rdmacm_cm_ep.c#L153

Reply:
1. can't we run out of qp numbers eventually?
2. what is the purpose to have a bitmap then? just save next_avail_qpn offset

Reply:
1. As MarkZhang told that even if the qpn was created and destoyed, the FW won't provide it again when SW creating a new qpn block again. Maybe after all qpn blocks were used, then FW will provide some destoyed(freed) qpn blk again.
2. bitmap also store status of all qpn in current block, for example if all 0 and all bits were used, then can destroy the block.

Reply:
2. can we keep a reference counter instead?

Reply:
> 2. can we keep a reference counter instead?
@yosefe Currently I can see only one thing against this: we need the bitmap to record the current status of the qpn slot(valid/invalid). If it's valid(0), then it means the qpn was destoryed or not created. We should avoid to destory it. But seems like we also have flag UCT_RDMACM_CM_EP_QPN_CREATED in upper layer can cover this.
So it should be fine if we want to remove this mechanism in the low level and only depend on the high level checking. WDYT?

Reply:
if we assume that code is creating/releasing everything correctly, no need the low level check

Reply:
@yosefe Fixed.